SlideBook 6 reader

RichardMyers opened the pull request:
First version of new SlideBook6 reader plugin.
- this reader is based on a JNI DLL wrapper that calls SlideBook 6 file format reader DLL
- reader includes simple meta data support
- DLLS are for Windows and x64 only

To work, the following .dll files must be in the ImageJ bin path (eg. jre/bin):
  SlideBook6Reader.dll  [x32](https://drive.google.com/open?id=0B5XYovwfrhQDfnRlTHBJUURJbHdBREJuZGJlYUVHdnN3Z1NzSUthMm0zWWZ5WEpPQl93a1E&authuser=0) [x64](https://drive.google.com/open?id=0B5XYovwfrhQDflN2QU1INUwzRC1pMVA3WHlyWExiUjQ2bWptcGIyLTZTXy1jZ250Uzc2RDA&authuser=0)
  SBReadFile.dll  [x32](https://drive.google.com/open?id=0B5XYovwfrhQDfnRlTHBJUURJbHdBREJuZGJlYUVHdnN3Z1NzSUthMm0zWWZ5WEpPQl93a1E&authuser=0)  [x64](https://drive.google.com/open?id=0B5XYovwfrhQDflN2QU1INUwzRC1pMVA3WHlyWExiUjQ2bWptcGIyLTZTXy1jZ250Uzc2RDA&authuser=0) 

Should I put the two required DLLs in the bioformats source tree? 

They will also be distributed as part of our free [SlideBook Reader](https://www.intelligent-imaging.com/reader.php) application.

Questions? comments?

melissalinkert commented:
Thank you for this, @RichardMyers.

I'll test this more thoroughly on Windows tomorrow, but for the moment could you please remove the comments that are directly copied from the reader guide documentation?

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,402 @@
+			int numPositions = getNumPositions(capture);
+			int numTimepoints = getNumTimepoints(capture);
+			int numChannels = getNumChannels(capture);
+			int numZPlanes = getNumZPlanes(capture);
```
It might be a little faster to use the ```sizeZ```, ```sizeT```, ```sizeC``` values from CoreMetadata (as set above) instead of making additional native method calls here.

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,402 @@
+
+		try {
+			openFile(id);
+			int numSeries = getNumCaptures();
```
Is it possible to have multiple positions or montage tiles per capture?  If so, ```numSeries``` should likely reflect the sum of all positions and montage tiles across all captures - otherwise only the first position from each tile will be accessible.

RichardMyers replied:
I see what you mean; is it possible to have dimensionOrder="XYCZPT" ?  When we capture a montage we capture all channels and planes for each montage xy stage position and may perform that loop for multiple time points.

joshmoore commented:
Closing & re-opening to adjust travis status.

RichardMyers commented:
I have updated the dll links and plugin to handle montages properly.   Comments are cleaned now too; let me know if you think it needs anything else.  I think the next step is to get some sample data for testing.

Montages positions are handled as time points in a single series, but each position of a multipoint capture that is not a montage is a new series.  Does that sound correct to you?  This matches how we treat these situations in SlideBook.

melissalinkert commented:
That looks much better, thanks.  There is one issue with the close() method, which I'll note separately, but this reads OK now.

We have plenty of test data already - as soon as I have access to the latest .dlls, I'll test here and verify that the images and generated OME-XML look fine.  The handling of montage positions should be fine, particularly if it matches what SlideBook users are used to seeing.

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,393 @@
+	// -- Internal FormatReader API methods --
+	public void close(boolean fileOnly) throws IOException {
+		super.close(fileOnly);
+		closeFile();
```
The call to ```closeFile()``` needs to be in an ```if (libraryFound)``` block.  ```close(fileOnly)``` can be called on an uninitialized reader, which in this case means that an ```UnsatisfiedLinkError``` will be thrown when the .dlls are not present even if this reader isn't being used.

RichardMyers commented:
Thanks @melissalinkert  I've updated the sharing on the links so anyone can access them.  I also fixed close(fileOnly) - good catch.

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,395 @@
+
+			// fill in meta data
+			MetadataStore store = makeFilterMetadata();
+			MetadataTools.populatePixels(store, this);
```
This should be ```MetadataTools.populatePixels(store, this, true);```, since attributes on Plane are being populated later on.

melissalinkert replied:
This line will still need to be updated before we can include the pull request - otherwise, all OME-XML generated via this reader will be invalid.

RichardMyers replied:
Thanks for the heads up; I've added the missing parameter.

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,395 @@
+					// link Objective to Image
+					String objectiveID =
+							MetadataTools.createLSID("Objective", 0, objectiveIndex);
+					store.setObjectiveID(objectiveID, 0, objectiveIndex);
```
```setObjectiveID``` should be called before the other ```setObjective*``` methods.

melissalinkert commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,395 @@
+			int[] numTimepoints = new int[numCaptures];
+			int[] numZPlanes = new int[numCaptures];
+			int[] numChannels = new int[numCaptures];
+			int[] numImages = new int[numCaptures];
```
```numImages``` can be removed, as it's no longer used.

melissalinkert commented:
Aside from the few minor comments above, the only worry I have is that there seems to be a lot (~4000 lines) of debugging output during setId, of the form:

```
CMaskDataSharedPtr()
operator=()
Uncompress
CMaskDataSharedPtr(const CMaskDataSharedPtr& inCopy)
```

I assume that's coming from SlideBook6Reader.dll, but if there's any way to remove that or turn it off, that would be helpful.

Otherwise, this looks good - once the comments above are addressed, this should be fine to include for the CI builds.

RichardMyers commented:
I found the offending cout << messages filling up the debugging output log.  I removed them and updated the links to new dlls.  I also compiled x32 dlls but haven't had a chance to test them yet.
I also fixed the setObjective  call order so hopefully everything is in good shape for wider testing.

RichardMyers commented:
@melissalinkert thanks for the heads up; I've added that missing parameter to populatePixels now.

sbesson commented on components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java:
```diff
@@ -0,0 +1,388 @@
+	}
+
+	/**
+	 * @see IFormatReader#isThisType(String, boolean)
```
This `@see` reference is causing a warning in the Javadocs (see https://ci.openmicroscopy.org/job/BIOFORMATS-5.1-merge-build/861/warnings39Result/). You may to use

```
@see loci.formats.IFormatReader#isThisType(String, boolean)
```

as above

melissalinkert commented:
Thanks, @RichardMyers.  Merging this so that it's sure to be in 5.1.2.

RichardMyers commented:
Thanks you @melissalinkert for walking me through the process.  I'm sure once this starts to get used I will have more pull requests based on the feedback we get.

ehrenfeu commented:
Does this mean only Windows x64 will be able to use Bio-Formats for reading SlideBook6 files?

RichardMyers commented:
I have now created both sets of dlls so both x32 and x64 versions should
work fine.
On May 28, 2015 13:53, "Niko Ehrenfeuchter" <notifications@github.com>
wrote:

> Does this mean only Windows x64 will be able to use Bio-Formats for
> reading SlideBook6 files?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openmicroscopy/bioformats/pull/1755#issuecomment-106288789>
> .
>

ehrenfeu commented:
Thanks, @RichardMyers ! I'm actually less worried about x32 personally, but more about Mac and Linux - my question was rather if the behaviour of Bio-Formats will be different on these platforms...?

joshmoore commented:
On any platform that the shared library ([`System.loadLibrary("SlideBook6Reader")`](https://github.com/openmicroscopy/bioformats/blob/4d332dd5f71d7ce56b35e16967668eaad1dd4f2b/components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java#L88)) can be successfully loaded, then the logic here will be used. Otherwise, Bio-Formats will fall back to using the previous reader ([`loci.formats.in.SlidebookReader       # sld`](https://github.com/openmicroscopy/bioformats/blob/4d332dd5f71d7ce56b35e16967668eaad1dd4f2b/components/formats-api/src/loci/formats/readers.txt#L39)).

ehrenfeu commented:
Thanks, Josh! So the `System.loadLibaray()` call will only succeed on Windows, as that's the only platform the DLLs are provided for, right?

joshmoore commented:
...will only succeed where you've correctly installed those libraries. (They don't ship with Bio-Formats itself)

RichardMyers commented:
That's right, we only have Windows DLLs built up now.  We have done some
work on an OS X version and so that is our next goal.  We don't think Linux
DLLs will be too different from OS X.

Thanks, Josh! So the System.loadLibaray() call will only succeed on
Windows, as that's the only platform the DLLs are provided for, right?

—
Reply to this email directly or view it on GitHub
<https://github.com/openmicroscopy/bioformats/pull/1755#issuecomment-106306014>
.

ehrenfeu commented:
Thanks all for the clarifications!
